### PR TITLE
CB-6644 Auditing environment

### PR DESCRIPTION
**Pull request by topolyai5**
The structured event moved to a different module, named as 'structuredevent-service-legacy'. In this modul the functionalities are untouched. For the CDP world, we created two new modules: 'structuredevent-api-cdp' and 'structuredevent-service-cdp'. We created a common endpoint, service and repository. Only one thing need to integrate: the domain SQL.

### CODE REVIEW

File: authorization-common/src/main/java/com/sequenceiq/authorization/service/ResourcePermissionChecker.java [hunk 1]
```diff
@@ -15,6 +16,9 @@
 
     protected Map<String, AuthorizationResourceAction> getAuthorizationActions(String resourceCrn, AuthorizationResourceAction action) {
         ResourceBasedCrnProvider resourceBasedCrnProvider = commonPermissionCheckingUtils.getResourceBasedCrnProvider(action);
+        if (resourceBasedCrnProvider == null) {
+            return Collections.emptyMap();
+        }
```
Comment (lacikaaa) [inline authorization-common/src/main/java/com/sequenceiq/authorization/service/ResourcePermissionChecker.java:21]: could you add an else branch or restructure the code so you won't have a return in the middle of the method?

File: common/src/main/java/com/sequenceiq/cloudbreak/common/json/Json.java [hunk 2]
```diff
@@ -170,4 +178,28 @@ public String toString() {
...
+            return false;
+        }
+    }
+
+    public boolean isArray() {
+        try {
+            JSONArray.fromObject(value);
+            return true;
+        } catch (JSONException e) {
+            LOGGER.debug("This json is not an object: {}", e.getMessage());
```
Comment (lacikaaa) [inline common/src/main/java/com/sequenceiq/cloudbreak/common/json/Json.java:197]: object->array

File: core/src/main/java/com/sequenceiq/cloudbreak/structuredevent/BaseLegacyStructuredFlowEventFactory.java [hunk 3]
```diff
@@ -26,18 +26,18 @@
...
 import com.sequenceiq.cloudbreak.structuredevent.event.StackDetails;
 import com.sequenceiq.cloudbreak.structuredevent.event.StructuredFlowEvent;
 import com.sequenceiq.cloudbreak.structuredevent.event.StructuredNotificationEvent;
-import com.sequenceiq.cloudbreak.structuredevent.rest.StructuredFlowEventFactory;
+import com.sequenceiq.cloudbreak.structuredevent.event.legacy.OperationDetails;
+import com.sequenceiq.cloudbreak.structuredevent.rest.LegacyStructuredFlowEventFactory;
 import com.sequenceiq.flow.ha.NodeConfig;
 
 @Component
 @Transactional
```
Comment (lacikaaa) [inline core/src/main/java/com/sequenceiq/cloudbreak/structuredevent/BaseLegacyStructuredFlowEventFactory.java:37]: could you check if we really need this transactional annotation?

File: environment/src/main/java/com/sequenceiq/environment/environment/domain/Environment.java [hunk 4]
```diff
@@ -156,6 +157,11 @@ public void setAccountId(String accountId) {
         this.accountId = accountId;
     }
 
+    @Override
+    public String getResourceName() {
+        return null;
```
Comment (lacikaaa) [inline environment/src/main/java/com/sequenceiq/environment/environment/domain/Environment.java:162]: why null?

File: environment/src/main/java/com/sequenceiq/environment/events/CredentialUrlParser.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+import java.util.regex.Pattern;
+
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.stereotype.Component;
+
+import com.sequenceiq.cloudbreak.structuredevent.event.CloudbreakEventService;
+import com.sequenceiq.cloudbreak.structuredevent.rest.urlparser.CDPRestUrlParser;
+
+@Component
+public class CredentialUrlParser extends CDPRestUrlParser {
```
Comment (lacikaaa) [inline environment/src/main/java/com/sequenceiq/environment/events/CredentialUrlParser.java:13]: missing unit test, especially for pattern matching

File: environment/src/main/java/com/sequenceiq/environment/events/EnvironmentStructuredFlowEventFactory.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    @Inject
+    private NodeConfig nodeConfig;
+
+    @Value("${info.app.version:}")
+    private String serviceVersion;
+
+    @Override
+    public CDPStructuredFlowEvent createStructuredFlowEvent(Long resourceId, FlowDetails flowDetails, Boolean detailed) {
+        Optional<Environment> environmentOptional = environmentService.findEnvironmentById(resourceId);
+        Environment environment = environmentOptional.orElseThrow(notFound("Environment", resourceId));
```
Comment (lacikaaa) [inline environment/src/main/java/com/sequenceiq/environment/events/EnvironmentStructuredFlowEventFactory.java:41]: I would prefer if you would add an `getEnvironmentById` method to the `EnvironmentService` and that could throw an exception as I see this kind of usage a lot in the code generally

File: environment/src/main/java/com/sequenceiq/environment/events/EnvironmentStructuredFlowEventFactory.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+    @Override
+    public CDPStructuredFlowEvent createStructuredFlowEvent(Long resourceId, FlowDetails flowDetails, Boolean detailed, Exception exception) {
+        return null;
+    }
+
+    @Override
+    public CDPStructuredNotificationEvent createStructuredNotificationEvent(Long resourceId, String notificationType, String message, String groupName) {
+        return null;
+    }
```
Comment (lacikaaa) [inline environment/src/main/java/com/sequenceiq/environment/events/EnvironmentStructuredFlowEventFactory.java:57]: I see you have created `CDPStructuredFlowEventFactory` and this is the only implementation of that interface. And 2 of 3 methods return null. Why do we need these methods?

File: environment/src/main/java/com/sequenceiq/environment/events/EventSenderService.java [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+
+    private CDPStructuredNotificationDetails getNotificationDetails(ResourceEvent resourceEvent, String resourceCrn, String resourceType,
+            Object payload) {
+        String serializedPayload;
+        try {
+            serializedPayload = new Gson().toJson(payload);
+            LOGGER.debug("CDPStructuredNotificationDetails' payload has been serialized with ResourceEvent[{}], resource type[{}], CRN[{}]",
+                    resourceEvent.name(), resourceType, resourceCrn);
+        } catch (RuntimeException re) {
+            String msg = String.format("CDPStructuredNotificationDetails' payload couldn't be serialized with ResourceEvent[%s], resource type[%s], CRN[%s]",
```
Comment (lacikaaa) [inline environment/src/main/java/com/sequenceiq/environment/events/EventSenderService.java:117]: I see no reason to introduce `msg` here, you could just use `serializedPayload`

File: structuredevent-api-cdp/src/main/java/com/sequenceiq/cloudbreak/structuredevent/rest/annotation/TransformGetterType.java [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Inherited;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Inherited
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.FIELD)
+public @interface TransformGetterType {
```
Comment (lacikaaa) [inline structuredevent-api-cdp/src/main/java/com/sequenceiq/cloudbreak/structuredevent/rest/annotation/TransformGetterType.java:12]: there is already a `TransformGetterType` in the codebase, is it possible to reuse that one? same applies for `Immutable`

File: structuredevent-api-cdp/src/main/java/com/sequenceiq/cloudbreak/structuredevent/rest/model/CDPEventV1Response.java [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonInclude.Include;
+import com.sequenceiq.cloudbreak.structuredevent.event.LdapDetails;
+import com.sequenceiq.cloudbreak.structuredevent.event.RdsDetails;
+
+import io.swagger.annotations.ApiModel;
+
+@ApiModel
+@JsonInclude(Include.NON_NULL)
+public class CDPEventV1Response extends CDPEventBaseV1 {
```
Comment (lacikaaa) [inline structuredevent-api-cdp/src/main/java/com/sequenceiq/cloudbreak/structuredevent/rest/model/CDPEventV1Response.java:12]: maybe add json ignore unknown annotation also

File: structuredevent-model/src/main/java/com/sequenceiq/cloudbreak/structuredevent/event/cdp/CDPOperationDetails.java [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+
+    public Long getTimestamp() {
+        return timestamp;
+    }
+
+    public void setTimestamp(Long timestamp) {
+        this.timestamp = timestamp;
+    }
+
+    public String getUTCDateTime() {
```
Comment (lacikaaa) [inline structuredevent-model/src/main/java/com/sequenceiq/cloudbreak/structuredevent/event/cdp/CDPOperationDetails.java:104]: this might worth a simple unit test

File: structuredevent-model/src/main/java/com/sequenceiq/cloudbreak/structuredevent/event/cdp/CDPStructuredEventContainer.java [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+        List<CDPStructuredFlowEvent> flow,
+        List<CDPStructuredRestCallEvent> rest,
+        List<CDPStructuredNotificationEvent> notification) {
+        this.flow = flow;
+        this.rest = rest;
+        this.notification = notification;
+    }
+
+    public List<CDPStructuredFlowEvent> getFlow() {
+        return flow != null ? flow : Collections.emptyList();
```
Comment (lacikaaa) [inline structuredevent-model/src/main/java/com/sequenceiq/cloudbreak/structuredevent/event/cdp/CDPStructuredEventContainer.java:30]: why not initialize the field with a `Collections.emptyList()` and then the getter won't have any logic. At least I think we usually handle this way. same applies to the others

File: structuredevent-model/src/main/java/com/sequenceiq/cloudbreak/structuredevent/event/cdp/CDPStructuredFlowEvent.java [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+        String state = flow.getFlowState();
+//        if ("unknown".equals(state)) {
+//            state = getOperation().getResourceType().toUpperCase()
+//            if (stack != null) {
+//                state = "STACK_" + stack.getDetailedStatus();
+//            }
+//            if (cluster != null) {
+//                state += " | CLUSTER_" + cluster.getStatus();
+//            }
+//        }
```
Comment (lacikaaa) [inline structuredevent-model/src/main/java/com/sequenceiq/cloudbreak/structuredevent/event/cdp/CDPStructuredFlowEvent.java:41]: do we need these or could be deleted? if can be then you should return in the first line and you don't need the state variaable

File: structuredevent-model/src/main/java/com/sequenceiq/cloudbreak/structuredevent/event/cdp/CDPStructuredFlowEvent.java [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import com.sequenceiq.cloudbreak.structuredevent.event.FlowDetails;
+import com.sequenceiq.cloudbreak.structuredevent.json.Base64Deserializer;
+import com.sequenceiq.cloudbreak.structuredevent.json.Base64Serializer;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class CDPStructuredFlowEvent<T> extends CDPStructuredEvent {
```
Comment (lacikaaa) [inline structuredevent-model/src/main/java/com/sequenceiq/cloudbreak/structuredevent/event/cdp/CDPStructuredFlowEvent.java:11]: I think `T` should be serializable

File: structuredevent-model/src/main/java/com/sequenceiq/cloudbreak/structuredevent/event/cdp/CDPStructuredFlowEvent.java [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+//            if (cluster != null) {
+//                state += " | CLUSTER_" + cluster.getStatus();
+//            }
+//        }
+        return state;
+    }
+
+    @Override
+    public Long getDuration() {
+        return Math.max(0L, flow.getDuration());
```
Comment (lacikaaa) [inline structuredevent-model/src/main/java/com/sequenceiq/cloudbreak/structuredevent/event/cdp/CDPStructuredFlowEvent.java:47]: why do we need this max? duration can be negative?

File: structuredevent-model/src/main/java/com/sequenceiq/cloudbreak/structuredevent/event/cdp/CDPStructuredNotificationEvent.java [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+
+    @Override
+    public String getStatus() {
+        return "SENT";
+    }
+
+    @Override
+    public Long getDuration() {
+        return 0L;
+    }
```
Comment (lacikaaa) [inline structuredevent-model/src/main/java/com/sequenceiq/cloudbreak/structuredevent/event/cdp/CDPStructuredNotificationEvent.java:47]: these 2 could be static final variable

File: structuredevent-model/src/main/java/com/sequenceiq/cloudbreak/structuredevent/event/cdp/CDPStructuredNotificationEvent.java [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+
+    public CDPStructuredNotificationEvent() {
+        super(CDPStructuredNotificationEvent.class.getSimpleName());
+    }
+
+    public CDPStructuredNotificationEvent(CDPOperationDetails operation, CDPStructuredNotificationDetails notificationDetails) {
+        super(CDPStructuredNotificationEvent.class.getSimpleName(), operation);
+        this.notificationDetails = notificationDetails;
+        ldapNotificationDetails = null;
+        rdsNotificationDetails = null;
```
Comment (lacikaaa) [inline structuredevent-model/src/main/java/com/sequenceiq/cloudbreak/structuredevent/event/cdp/CDPStructuredNotificationEvent.java:24]: why these null initializations required? the same for those in the other constructors

File: structuredevent-model/src/main/java/com/sequenceiq/cloudbreak/structuredevent/event/cdp/CDPStructuredRestCallEvent.java [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    public CDPStructuredRestCallEvent(CDPOperationDetails operation, RestCallDetails restCall) {
+        super(CDPStructuredRestCallEvent.class.getSimpleName(), operation);
+        this.restCall = restCall;
+    }
+
+    @Override
+    public String getStatus() {
+        return restCall.getRestResponse().getStatusText() != null
```
Comment (lacikaaa) [inline structuredevent-model/src/main/java/com/sequenceiq/cloudbreak/structuredevent/event/cdp/CDPStructuredRestCallEvent.java:21]: are we sure restcall has always a non null restresponse? I don't see anything in `RestCallDetails` what would ensure that

File: structuredevent-service-cdp/src/main/java/com/sequenceiq/cloudbreak/structuredevent/config/CDPStructuredEventConfig.java [hunk 19]
```diff
@@ -0,0 +1,7 @@
+package com.sequenceiq.cloudbreak.structuredevent.config;
+
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+public class CDPStructuredEventConfig {
```
Comment (lacikaaa) [inline structuredevent-service-cdp/src/main/java/com/sequenceiq/cloudbreak/structuredevent/config/CDPStructuredEventConfig.java:6]: do we need this empty class?

File: structuredevent-service-cdp/src/main/java/com/sequenceiq/cloudbreak/structuredevent/rest/CDPRestCommonService.java [hunk 20]
```diff
@@ -0,0 +1,129 @@
...
+            }
+        }
+        if (responseJson != null && StringUtils.isEmpty(id)) {
+                id = Optional.ofNullable(getValueFromJson(responseJson, path))
+                        .orElse(getListValue(request, responseJson, pluralPath, restParams, idType));
+        }
+        return id;
+    }
+
+    private String getListValue(RestRequestDetails restRequest, Json json, String path, Map<String, String> restParams, String idType) {
```
Comment (lacikaaa) [inline structuredevent-service-cdp/src/main/java/com/sequenceiq/cloudbreak/structuredevent/rest/CDPRestCommonService.java:83]: it's a bit hard to follow this method, similarly to `getResourceId`. Could you refactor this a bit to make it more readable?

File: structuredevent-service-cdp/src/main/java/com/sequenceiq/cloudbreak/structuredevent/rest/CDPRestCommonService.java [hunk 21]
```diff
@@ -0,0 +1,129 @@
...
+import org.springframework.stereotype.Component;
+
+import com.sequenceiq.cloudbreak.common.anonymizer.AnonymizerUtil;
+import com.sequenceiq.cloudbreak.common.json.Json;
+import com.sequenceiq.cloudbreak.common.json.JsonUtil;
+import com.sequenceiq.cloudbreak.structuredevent.event.rest.RestCallDetails;
+import com.sequenceiq.cloudbreak.structuredevent.event.rest.RestRequestDetails;
+
+@Component
+public class CDPRestCommonService {
```
Comment (lacikaaa) [inline structuredevent-service-cdp/src/main/java/com/sequenceiq/cloudbreak/structuredevent/rest/CDPRestCommonService.java:22]: please add unit tests

File: structuredevent-service-cdp/src/main/java/com/sequenceiq/cloudbreak/structuredevent/rest/filter/CDPStructuredEventFilter.java [hunk 22]
```diff
@@ -0,0 +1,383 @@
...
+
+    private static final String CRN = "crn";
+
+    private final List<String> skippedHeadersList = Lists.newArrayList("authorization");
+
+    private final Map<String, AccountAwareResourceRepository<?, ?>> pathRepositoryMap = new HashMap<>();
+
+    private final Pattern extractIdRestParamFromResponsePattern = Pattern.compile("\"" + ID + "\":([0-9]*)");
+
+    private final Pattern extractCrnRestParamFromResponsePattern = Pattern.compile("\"" + CRN + "\":\"([0-9a-zA-Z:-]*)\"");
```
Comment (lacikaaa) [inline structuredevent-service-cdp/src/main/java/com/sequenceiq/cloudbreak/structuredevent/rest/filter/CDPStructuredEventFilter.java:97]: not used anywhere

File: structuredevent-service-cdp/src/main/java/com/sequenceiq/cloudbreak/structuredevent/rest/filter/CDPStructuredEventFilter.java [hunk 23]
```diff
@@ -0,0 +1,383 @@
...
+                    MessageUtils.getCharset(context.getMediaType())).toString();
+            Map<String, String> restParams = (Map<String, String>) context.getProperty(REST_PARAMS);
+            if (restParams == null) {
+                restParams = new HashMap<>();
+            }
+            sendStructuredEvent(restRequest, restResponse, restParams, requestTime, responseBody);
+        }
+    }
+
+    private void sendStructuredEvent(RestRequestDetails restRequest, RestResponseDetails restResponse, Map<String, String> restParams, Long requestTime,
```
Comment (lacikaaa) [inline structuredevent-service-cdp/src/main/java/com/sequenceiq/cloudbreak/structuredevent/rest/filter/CDPStructuredEventFilter.java:203]: could you break up this method? it too long and could be refactored

File: structuredevent-service-cdp/src/main/java/com/sequenceiq/cloudbreak/structuredevent/rest/filter/CDPStructuredEventFilter.java [hunk 24]
```diff
@@ -0,0 +1,383 @@
...
+            String responseBody) {
+        boolean valid = checkRestParams(restParams);
+        try {
+            if (!valid) {
+                LOGGER.debug("Cannot create structured event, because rest params are invalid.");
+                return;
+            }
```
Comment (lacikaaa) [inline structuredevent-service-cdp/src/main/java/com/sequenceiq/cloudbreak/structuredevent/rest/filter/CDPStructuredEventFilter.java:210]: please don't return in the middle of the method, at least add an `else` branch

File: structuredevent-service-cdp/src/main/java/com/sequenceiq/cloudbreak/structuredevent/rest/filter/CDPStructuredEventFilter.java [hunk 25]
```diff
@@ -0,0 +1,383 @@
...
+import com.sequenceiq.cloudbreak.structuredevent.rest.CDPRestCommonService;
+import com.sequenceiq.cloudbreak.structuredevent.rest.annotation.AccountEntityType;
+import com.sequenceiq.cloudbreak.structuredevent.rest.urlparser.CDPRestUrlParser;
+import com.sequenceiq.cloudbreak.structuredevent.service.CDPBaseRestRequestThreadLocalService;
+import com.sequenceiq.cloudbreak.structuredevent.service.CDPDefaultStructuredEventClient;
+import com.sequenceiq.cloudbreak.structuredevent.service.lookup.CDPAccountAwareRepositoryLookupService;
+import com.sequenceiq.flow.ha.NodeConfig;
+
+@Component
+public class CDPStructuredEventFilter implements WriterInterceptor, ContainerRequestFilter, ContainerResponseFilter {
```
Comment (lacikaaa) [inline structuredevent-service-cdp/src/main/java/com/sequenceiq/cloudbreak/structuredevent/rest/filter/CDPStructuredEventFilter.java:69]: this is a bit complicated class with a lot of logic and hard to follow/understand. no one would dare to touch it if you don't cover this with unit tests pretty well.
also some refactor would be nice, the high number of `Autowires` indicate the same also that maybe the whole class could be refactored into smaller ones which is easier to understand, maintain and test

File: structuredevent-service-cdp/src/main/java/com/sequenceiq/cloudbreak/structuredevent/service/AbstractAccountAwareResourceService.java [hunk 26]
```diff
@@ -0,0 +1,156 @@
...
+import com.sequenceiq.cloudbreak.structuredevent.repository.AccountAwareResource;
+import com.sequenceiq.cloudbreak.structuredevent.repository.AccountAwareResourceRepository;
+
+public abstract class AbstractAccountAwareResourceService<T extends AccountAwareResource>
+        implements AccountAwareResourceService<T> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(AbstractAccountAwareResourceService.class);
+
+    @Override
+    public T create(T resource, @Nonnull String accountId) {
```
Comment (lacikaaa) [inline structuredevent-service-cdp/src/main/java/com/sequenceiq/cloudbreak/structuredevent/service/AbstractAccountAwareResourceService.java:28]: maybe `@NotEmpty` would fit better. Should we add this to the other public methods definition also?

File: structuredevent-service-cdp/src/main/java/com/sequenceiq/cloudbreak/structuredevent/rest/filter/CDPStructuredEventFilter.java (hunk 25 above)
Comment (biharitomi) [reply structuredevent-service-cdp/src/main/java/com/sequenceiq/cloudbreak/structuredevent/rest/filter/CDPStructuredEventFilter.java:69]: Hi, yes, it makes sense, but this logic came from the core module. Good idea to refactor it, but it could be a separated task/improvement. Also my experience with this class that dependencies could not be injected into Jersey managed filters with `Inject` annotation and the only way that worked for me leave the `Autowire`s. I only could dig in a little bit, but in case of `Inject` Jersey tries to use it's own dependency management mechanism to resolve dependencies which are configured in Spring related configurations and managed by Spring.

File: structuredevent-service-cdp/src/main/java/com/sequenceiq/cloudbreak/structuredevent/service/CDPFlowStructuredEventHandler.java [hunk 27]
```diff
@@ -0,0 +1,133 @@
...
+import org.springframework.stereotype.Component;
+
+import com.sequenceiq.cloudbreak.structuredevent.event.FlowDetails;
+import com.sequenceiq.cloudbreak.structuredevent.event.cdp.CDPStructuredEvent;
+import com.sequenceiq.flow.core.FlowEventListener;
+
+@Primary
+@Component
+@Scope("prototype")
+public class CDPFlowStructuredEventHandler<S, E> extends StateMachineListenerAdapter<S, E> implements FlowEventListener<S, E> {
```
Comment (lacikaaa) [inline structuredevent-service-cdp/src/main/java/com/sequenceiq/cloudbreak/structuredevent/service/CDPFlowStructuredEventHandler.java:24]: please add unit tests

File: structuredevent-service-cdp/src/main/java/com/sequenceiq/cloudbreak/structuredevent/service/converter/CDPStructuredEventEntityToCDPStructuredEventConverter.java [hunk 28]
```diff
@@ -0,0 +1,53 @@
...
+import com.sequenceiq.cloudbreak.common.json.JsonUtil;
+import com.sequenceiq.cloudbreak.converter.AbstractConversionServiceAwareConverter;
+import com.sequenceiq.cloudbreak.structuredevent.domain.CDPStructuredEventEntity;
+import com.sequenceiq.cloudbreak.structuredevent.event.cdp.CDPStructuredEvent;
+import com.sequenceiq.cloudbreak.structuredevent.event.cdp.CDPStructuredFlowEvent;
+import com.sequenceiq.cloudbreak.structuredevent.event.cdp.CDPStructuredNotificationEvent;
+import com.sequenceiq.cloudbreak.structuredevent.event.cdp.CDPStructuredRestCallEvent;
+
+@Component
+public class CDPStructuredEventEntityToCDPStructuredEventConverter
```
Comment (lacikaaa) [inline structuredevent-service-cdp/src/main/java/com/sequenceiq/cloudbreak/structuredevent/service/converter/CDPStructuredEventEntityToCDPStructuredEventConverter.java:23]: unit test

File: structuredevent-service-cdp/src/main/java/com/sequenceiq/cloudbreak/structuredevent/service/converter/CDPStructuredEventToCDPStructuredEventEntityConverter.java [hunk 29]
```diff
@@ -0,0 +1,36 @@
...
+            structuredEventEntity.setStructuredEventJson(new Json(source));
+            CDPOperationDetails operationDetails = source.getOperation();
+            structuredEventEntity.setEventType(operationDetails.getEventType());
+            structuredEventEntity.setResourceType(operationDetails.getResourceType());
+            structuredEventEntity.setResourceCrn(operationDetails.getResourceCrn());
+            structuredEventEntity.setTimestamp(operationDetails.getTimestamp());
+            structuredEventEntity.setAccountId(source.getOperation().getAccountId());
+            return structuredEventEntity;
+        } catch (IllegalArgumentException e) {
+            LOGGER.error("Failed to parse structured event JSON", e);
```
Comment (lacikaaa) [inline structuredevent-service-cdp/src/main/java/com/sequenceiq/cloudbreak/structuredevent/service/converter/CDPStructuredEventToCDPStructuredEventEntityConverter.java:32]: shall we log the source to know what we can't convert?

File: structuredevent-service-cdp/src/main/java/com/sequenceiq/cloudbreak/structuredevent/service/lookup/CDPRepositoryLookupService.java [hunk 30]
```diff
@@ -0,0 +1,57 @@
...
+import javax.annotation.PostConstruct;
+
+import org.springframework.data.repository.CrudRepository;
+import org.springframework.util.CollectionUtils;
+
+import com.sequenceiq.cloudbreak.workspace.repository.EntityType;
+
+public abstract class CDPRepositoryLookupService<T extends CrudRepository<?, ?>> {
+
+    private List<T> repositoryList;
```
Comment (lacikaaa) [inline structuredevent-service-cdp/src/main/java/com/sequenceiq/cloudbreak/structuredevent/service/lookup/CDPRepositoryLookupService.java:17]: this field is unnecessary and misleading. could you remove this?

File: structuredevent-service-legacy/src/main/java/com/sequenceiq/cloudbreak/structuredevent/config/LegacyStructuredEventConfig.java [hunk 31]
```diff
@@ -0,0 +1,8 @@
+package com.sequenceiq.cloudbreak.structuredevent.config;
+
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+public class LegacyStructuredEventConfig {
```
Comment (lacikaaa) [inline structuredevent-service-legacy/src/main/java/com/sequenceiq/cloudbreak/structuredevent/config/LegacyStructuredEventConfig.java:6]: do we need this class?

File: structuredevent-service-cdp/src/main/java/com/sequenceiq/cloudbreak/structuredevent/rest/filter/CDPStructuredEventFilter.java (hunk 25 above)
Comment (lacikaaa) [reply structuredevent-service-cdp/src/main/java/com/sequenceiq/cloudbreak/structuredevent/rest/filter/CDPStructuredEventFilter.java:69]: I don't have any issues with `Autowire` :)
I'm fine with fixing this in a followup pr, just please create a jira for that, otherwise I'm afraid this will stay the same way

File: core/src/main/java/com/sequenceiq/cloudbreak/structuredevent/BaseLegacyStructuredFlowEventFactory.java (hunk 3 above)
Comment (topolyai5) [reply core/src/main/java/com/sequenceiq/cloudbreak/structuredevent/BaseLegacyStructuredFlowEventFactory.java:37]: Yes, it is necessary.

Comment (topolyai5) [reply core/src/main/java/com/sequenceiq/cloudbreak/structuredevent/BaseLegacyStructuredFlowEventFactory.java:37]: LazyInit exception occurred during the event firing

File: structuredevent-model/src/main/java/com/sequenceiq/cloudbreak/structuredevent/event/cdp/CDPStructuredRestCallEvent.java (hunk 18 above)
Comment (topolyai5) [reply structuredevent-model/src/main/java/com/sequenceiq/cloudbreak/structuredevent/event/cdp/CDPStructuredRestCallEvent.java:21]: yes, here: `CDPStructuredEventFilter#326-335`

File: structuredevent-service-cdp/src/main/java/com/sequenceiq/cloudbreak/structuredevent/rest/filter/CDPStructuredEventFilter.java (hunk 23 above)
Comment (topolyai5) [reply structuredevent-service-cdp/src/main/java/com/sequenceiq/cloudbreak/structuredevent/rest/filter/CDPStructuredEventFilter.java:203]: Here is the jira for tracking this: https://jira.cloudera.com/browse/CB-8777

File: structuredevent-service-cdp/src/main/java/com/sequenceiq/cloudbreak/structuredevent/rest/filter/CDPStructuredEventFilter.java (hunk 24 above)
Comment (topolyai5) [reply structuredevent-service-cdp/src/main/java/com/sequenceiq/cloudbreak/structuredevent/rest/filter/CDPStructuredEventFilter.java:210]: I will refactor this `return;` in the follow-up jira.

File: structuredevent-model/src/main/java/com/sequenceiq/cloudbreak/structuredevent/event/cdp/CDPStructuredFlowEvent.java (hunk 14 above)
Comment (topolyai5) [reply structuredevent-model/src/main/java/com/sequenceiq/cloudbreak/structuredevent/event/cdp/CDPStructuredFlowEvent.java:11]: I don't want to add the `Serializable` to the generic type. Sometimes generate unnecessary changes/refactors.

File: authorization-common/src/main/java/com/sequenceiq/authorization/service/ResourcePermissionChecker.java (hunk 1 above)
Comment (topolyai5) [reply authorization-common/src/main/java/com/sequenceiq/authorization/service/ResourcePermissionChecker.java:21]: I think it is much clearer then if I create an if {} else {} structure. Otherwise, we have a lot of similar code style in the codebase.

File: structuredevent-model/src/main/java/com/sequenceiq/cloudbreak/structuredevent/event/cdp/CDPStructuredFlowEvent.java (hunk 14 above)
Comment (lacikaaa) [reply structuredevent-model/src/main/java/com/sequenceiq/cloudbreak/structuredevent/event/cdp/CDPStructuredFlowEvent.java:11]: but `CDPStructuredEvent` is serializable and a class is only serializable if all the fields are serializable or transient. In this case you don't have to modify anything else as I have checked and the constructor is used only at one place where payload is set to `null`

File: structuredevent-service-cdp/src/main/java/com/sequenceiq/cloudbreak/structuredevent/service/CDPFlowStructuredEventHandler.java (hunk 27 above)
Comment (topolyai5) [reply structuredevent-service-cdp/src/main/java/com/sequenceiq/cloudbreak/structuredevent/service/CDPFlowStructuredEventHandler.java:24]: Hm, I tried to write any useful test, but I cannot. Could you please give me any suggestion?

Comment (lacikaaa) [reply structuredevent-service-cdp/src/main/java/com/sequenceiq/cloudbreak/structuredevent/service/CDPFlowStructuredEventHandler.java:24]: - in `public void transition(Transition<S, E> transition)` a `CDPStructuredEvent` is sent. in this meethod there are a lot of ternary operator and an if also. depending on the `transition` parameter one of the `cdpStructuredFlowEventFactory.createStructuredFlowEvent` is called with different values which could be asserted. 
you can even check the in case an error like nullpointer it is caught, so this would ensure no one removes the try catch by mistake.
- in `public void stateMachineStopped(StateMachine<S, E> stateMachine)` you could test it has only  effect if it's not complete. it also has some ternary operator, those effects could be tested also.

In the meanwhile: why do we have try-catch in `transition` and not in `stateMachineStopped`?

Comment (topolyai5) [reply structuredevent-service-cdp/src/main/java/com/sequenceiq/cloudbreak/structuredevent/service/CDPFlowStructuredEventHandler.java:24]: _In the meanwhile: why do we have try-catch in transition and not in stateMachineStopped?_
Good questions, these codes are 2-3 years old.

Comment (topolyai5) [reply structuredevent-service-cdp/src/main/java/com/sequenceiq/cloudbreak/structuredevent/service/CDPFlowStructuredEventHandler.java:24]: I don't agree with you. These tests could not catch any error. I cannot validate the content of the payload. It would be much useful than verifying the methods. 
But I will write these tests, maybe

**Review by lacikaaa — APPROVED** [review]
